File: lib/eie-scoring.js
// ─────────────────────────────────────────────────────────────────────────
// EIE Scoring — Phase 1 (deterministic aggregation over anchored ratings).
// This is intentionally the "degenerate case" of EVALUATION_ENGINE.md §6:
// one rater, linear level→score, no GRM yet. It already delivers the honest
// upgrades: Unknown/abstain, coverage, evidence-linked scores, mechanical
// composite, and utility-free banding. Bayesian θ + G-theory arrive in Phase 2/3.
// ─────────────────────────────────────────────────────────────────────────

import { grmPosterior, grmPosteriorMulti, compositePosterior, priorMeanForSeniority, normalCdf, raterAgreement, profileDependability } from './psychometrics.js'
import { decideHire } from './decision-engine.js'

const clamp = (n, lo, hi) => Math.max(lo, Math.min(hi, n))

// Merge the model (competency defs + weights) with the model-produced ratings
// into a finalized, defensible competency profile + decision.
export function finalizeProfile(model, ratings) {
  const byId = {}
  for (const r of ratings || []) if (r && r.competency_id) byId[r.competency_id] = r

  const priorMean = priorMeanForSeniority(model.seniority)

  const competencies = model.competencies.map((c) => {
    const r = byId[c.id] || {}
    const level = Number(r.bars_level)
    let confidence = clamp(Number(r.confidence) || 0, 0, 1)
    const minEp = c.evidence_requirements?.min_episodes ?? 1
    const episodes = Array.isArray(r.evidence) ? r.evidence.length : (r.evidence ? 1 : 0)

    // Abstain → "unknown" (NOT a low score) only when there is genuinely too
    // little information: no level, the rater abstained, or confidence is very
    // low. Falling short of the ideal episode count does NOT discard the signal —
    // it just widens uncertainty (confidence is capped). See §4.5.
    const insufficient = !Number.isFinite(level) || r.state === 'unknown' || confidence < 0.35
    if (!insufficient && episodes < minEp) confidence = Math.min(confidence, 0.55)
    const state = insufficient ? 'unknown' : 'measured'
    const thin = state === 'measured' && episodes < minEp
    const failure = Array.isArray(r.failure_triggered) ? r.failure_triggered : (r.failure_triggered ? [r.failure_triggered] : [])

    // ── Measurement: Bayesian posterior over latent ability θ (Phase 2) ──
    let post = null
    if (state === 'measured') {
      // Effective information = rater confidence × number of cited episodes
      // (each episode is a locally-independent observation of θ). More, better-
      // evidenced behavior ⇒ tighter credible interval & higher reliability.
      const effObs = confidence * clamp(episodes, 1, 3)
      post = grmPosterior({
        level: clamp(level, 1, 5),
        weight: effObs,
        a: c.discrimination,           // optional per-competency override
        priorMean,
      })
    }

    return {
      id: c.id, name: c.name, cluster: c.cluster, weight: c.weight,
      state, bars_level: state === 'measured' ? clamp(level, 1, 5) : null,
      // score (0–100) is the norm-population percentile of θ — intuitive UI metric.
      score: post ? post.percentile : null,
      theta: post ? post.theta : null,
      se: post ? post.se : null,
      ci90: post ? post.ci90 : null,
      ci_pct: post ? [Math.round(normalCdf(post.ci90[0]) * 100), Math.round(normalCdf(post.ci90[1]) * 100)] : null,
      reliability: post ? post.reliability : null,
      confidence,
      evidence: Array.isArray(r.evidence) ? r.evidence.slice(0, 3) : [],
      followup: state === 'unknown' ? (r.followup || defaultFollowup(c)) : null,
      thin,
      risk: failure,
    }
  })

  // Mechanical composite over MEASURED competencies only; unknowns excluded,
  // never imputed as average. Uncertainty propagated correctly (§4). coverage =
  // fraction of decision weight measured.
  const measured = competencies.filter((c) => c.state === 'measured')
  const comp = compositePosterior(measured.map((c) => ({ theta: c.theta, se: c.se, weight: c.weight })))
  const covered = measured.reduce((s, c) => s + c.weight, 0)
  const totalW = competencies.reduce((s, c) => s + c.weight, 0)
  const wMeasured = covered || 0

  const composite = comp ? comp.percentile : null      // 0–100 (percentile of composite θ)
  const coverage = totalW > 0 ? Math.round((covered / totalW) * 100) / 100 : 0
  const overallConfidence = wMeasured > 0
    ? Math.round((measured.reduce((s, c) => s + c.confidence * c.weight, 0) / wMeasured) * 100) / 100
    : 0
  const overallReliability = wMeasured > 0
    ? Math.round((measured.reduce((s, c) => s + (c.reliability || 0) * c.weight, 0) / wMeasured) * 100) / 100
    : 0

  const criticalFailures = competencies.filter((c) => c.risk && c.risk.length > 0).map((c) => c.name)
  const decision = decideHire({ theta: comp ? comp.theta : null, se: comp ? comp.se : null, coverage, criticalFailures, seniority: model.seniority })
  if (decision.band === 'Hire with Reservations' || decision.band === 'Borderline') {
    decision.reservations = competencies.filter((c) => c.state === 'measured' && c.weight >= 0.08 && c.score < 50).map((c) => c.name)
  }

  return {
    framework_version: model.framework_version,
    role_family: model.role_family,
    seniority: model.seniority,
    measurement_model: 'grm-eap-v1',
    competencies,
    composite_score: composite,        // 0–100 percentile of composite θ
    composite_theta: comp ? comp.theta : null,
    composite_ci90: comp ? comp.ci90 : null,
    coverage,                          // 0–1
    overall_confidence: overallConfidence,
    overall_reliability: overallReliability,
    decision,
    unknown_competencies: competencies.filter((c) => c.state === 'unknown').map((c) => c.name),
    followups: competencies.filter((c) => c.followup).map((c) => ({ competency: c.name, question: c.followup })),
  }
}


// ── Phase 3: ensemble aggregation ─────────────────────────────────────────
// byComp[competency_id] = array of per-rater ratings ({state, bars_level,
// confidence, evidence, failure_triggered, followup}). refutations[id] =
// { refuted, reason } from the adversarial pass.
export function finalizeProfileEnsemble(model, byComp, refutations = {}, meta = {}) {
  const priorMean = priorMeanForSeniority(model.seniority)
  const nRaters = meta.rater_count || 1

  const competencies = model.competencies.map((c) => {
    const raters = Array.isArray(byComp[c.id]) ? byComp[c.id] : []
    const measuredRaters = raters.filter((r) => r && r.state !== 'unknown' && Number.isFinite(Number(r.bars_level)) && (Number(r.confidence) || 0) >= 0.35)
    const minEp = c.evidence_requirements?.min_episodes ?? 1
    const refuted = !!refutations[c.id]?.refuted

    // Unknown if a MAJORITY of raters abstained / had too little.
    if (measuredRaters.length === 0 || measuredRaters.length < Math.ceil(raters.length / 2)) {
      const fu = raters.find((r) => r?.followup)?.followup || defaultFollowup(c)
      return { id: c.id, name: c.name, cluster: c.cluster, weight: c.weight, state: 'unknown', followup: fu, rater_count: raters.length }
    }

    const levels = measuredRaters.map((r) => clamp(Number(r.bars_level), 1, 5))
    const agreement = raterAgreement(levels)
    const episodesEach = measuredRaters.map((r) => (Array.isArray(r.evidence) ? r.evidence.length : (r.evidence ? 1 : 0)))
    const totalEpisodes = Math.max(...episodesEach, 1)
    const thin = totalEpisodes < minEp
    const refuteFactor = refuted ? 0.4 : 1

    // Each rater contributes one observation; weight = its confidence × its
    // cited episodes × refutation factor. Agreement/disagreement flows through
    // the multi-observation posterior automatically.
    const observations = measuredRaters.map((r, i) => {
      let conf = clamp(Number(r.confidence) || 0, 0, 1)
      if (thin) conf = Math.min(conf, 0.55)
      return { level: clamp(Number(r.bars_level), 1, 5), weight: conf * clamp(episodesEach[i], 1, 3) * refuteFactor }
    })

    const post = grmPosteriorMulti({ observations, a: c.discrimination, priorMean })
    const meanLevel = Math.round(levels.reduce((s, x) => s + x, 0) / levels.length)
    const meanConf = clamp(measuredRaters.reduce((s, r) => s + (Number(r.confidence) || 0), 0) / measuredRaters.length, 0, 1) * refuteFactor
    // best evidence = from the rater who cited the most
    const richest = measuredRaters.slice().sort((a, b) => (b.evidence?.length || 0) - (a.evidence?.length || 0))[0]
    const failures = measuredRaters.flatMap((r) => Array.isArray(r.failure_triggered) ? r.failure_triggered : (r.failure_triggered ? [r.failure_triggered] : []))

    return {
      id: c.id, name: c.name, cluster: c.cluster, weight: c.weight,
      state: 'measured', bars_level: meanLevel,
      score: post.percentile, theta: post.theta, se: post.se, ci90: post.ci90,
      ci_pct: [Math.round(normalCdf(post.ci90[0]) * 100), Math.round(normalCdf(post.ci90[1]) * 100)],
      reliability: post.reliability,
      confidence: Math.round(meanConf * 100) / 100,
      rater_count: measuredRaters.length,
      rater_agreement: agreement,
      refuted, refute_reason: refuted ? refutations[c.id]?.reason : null,
      evidence: Array.isArray(richest?.evidence) ? richest.evidence.slice(0, 3) : [],
      thin,
      risk: failures,
      _levels: levels,
    }
  })

  return assembleProfile(model, competencies, { ...meta, rater_count: nRaters })
}

// Shared assembly used by single-rater and ensemble paths.
function assembleProfile(model, competencies, meta = {}) {
  const measured = competencies.filter((c) => c.state === 'measured')
  const comp = compositePosterior(measured.map((c) => ({ theta: c.theta, se: c.se, weight: c.weight })))
  const covered = measured.reduce((s, c) => s + c.weight, 0)
  const totalW = competencies.reduce((s, c) => s + c.weight, 0)
  const wMeasured = covered || 0

  const composite = comp ? comp.percentile : null
  const coverage = totalW > 0 ? Math.round((covered / totalW) * 100) / 100 : 0
  const overallConfidence = wMeasured > 0
    ? Math.round((measured.reduce((s, c) => s + c.confidence * c.weight, 0) / wMeasured) * 100) / 100 : 0
  const overallReliability = wMeasured > 0
    ? Math.round((measured.reduce((s, c) => s + (c.reliability || 0) * c.weight, 0) / wMeasured) * 100) / 100 : 0

  // G-theory dependability of the profile against rater swaps (needs ≥2 raters
  // measuring ≥2 shared competencies). Build a RECTANGULAR competency×rater
  // matrix from competencies rated by the same number of raters.
  let dependability = null
  if ((meta.rater_count || 1) >= 2) {
    const withLevels = measured.filter((c) => Array.isArray(c._levels) && c._levels.length >= 2)
    const lenCounts = {}
    withLevels.forEach((c) => { lenCounts[c._levels.length] = (lenCounts[c._levels.length] || 0) + 1 })
    const L = Object.keys(lenCounts).map(Number).sort((a, b) => lenCounts[b] - lenCounts[a])[0]
    const matrix = withLevels.filter((c) => c._levels.length === L).map((c) => c._levels)
    dependability = profileDependability(matrix)
  }
  // strip internal field from output
  competencies.forEach((c) => { delete c._levels })

  const criticalFailures = competencies.filter((c) => c.risk && c.risk.length > 0).map((c) => c.name)
  const decision = decideHire({ theta: comp ? comp.theta : null, se: comp ? comp.se : null, coverage, criticalFailures, seniority: model.seniority })
  if (decision.band === 'Hire with Reservations' || decision.band === 'Borderline') {
    decision.reservations = competencies.filter((c) => c.state === 'measured' && c.weight >= 0.08 && c.score < 50).map((c) => c.name)
  }

  return {
    framework_version: model.framework_version,
    role_family: model.role_family,
    seniority: model.seniority,
    measurement_model: 'grm-eap-v1',
    rater_count: meta.rater_count || 1,
    rater_dependability: dependability?.dependability ?? null,
    variance_components: dependability ? { comp: dependability.vComp, rater: dependability.vRater, resid: dependability.vResid } : null,
    competencies,
    composite_score: composite,
    composite_theta: comp ? comp.theta : null,
    composite_ci90: comp ? comp.ci90 : null,
    coverage,
    overall_confidence: overallConfidence,
    overall_reliability: overallReliability,
    decision,
    unknown_competencies: competencies.filter((c) => c.state === 'unknown').map((c) => c.name),
    followups: competencies.filter((c) => c.followup).map((c) => ({ competency: c.name, question: c.followup })),
  }
}

function defaultFollowup(c) {
  return `Ask for a specific real example that would demonstrate ${c.name.toLowerCase()}.`
}

export const DECISION_BANDS = [
  'Strong Hire', 'Hire', 'Hire with Reservations', 'Borderline',
  'Needs More Evidence', 'No Hire', 'Strong No Hire',
]

File: backend/domain/evaluation/eie_scoring.py
"""EIE Scoring — deterministic aggregation over anchored ratings
(ported from lib/eie-scoring.js).

Merges the competency model (defs + weights) with model-produced BARS ratings
into a finalized, defensible profile + decision. Delivers the honest upgrades:
Unknown/abstain (never a fake low score), coverage, evidence-linked scores, a
mechanical uncertainty-propagated composite, and utility-derived banding.
Single-rater and N-rater ensemble paths share one assembler.
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Any

from domain.evaluation.decision_engine import Decision, decide_hire
from domain.evaluation.psychometrics import (
    Posterior,
    composite_posterior,
    grm_posterior,
    grm_posterior_multi,
    normal_cdf,
    prior_mean_for_seniority,
    profile_dependability,
    rater_agreement,
)

DECISION_BANDS = [
    "Strong Hire", "Hire", "Hire with Reservations", "Borderline",
    "Needs More Evidence", "No Hire", "Strong No Hire",
]


def _clamp(n: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, n))


def _js_round2(n: float) -> float:
    import math
    return math.floor(n * 100 + 0.5) / 100


def _default_followup(c: dict[str, Any]) -> str:
    return f"Ask for a specific real example that would demonstrate {c['name'].lower()}."


def _episodes(evidence: Any) -> int:
    if isinstance(evidence, list):
        return len(evidence)
    return 1 if evidence else 0


def _as_list(x: Any) -> list:
    if isinstance(x, list):
        return x
    return [x] if x else []


def _ci_pct(post: Posterior) -> list[int]:
    from domain.evaluation.psychometrics import _js_round
    return [_js_round(normal_cdf(post.ci90[0]) * 100), _js_round(normal_cdf(post.ci90[1]) * 100)]


def _decision_to_dict(d: Decision) -> dict[str, Any]:
    # Match the JS decision object shape: omit fields that are None (the
    # unknown/critical paths don't carry percentile/meets_bar/coverage).
    return {k: v for k, v in asdict(d).items() if v is not None}


def finalize_profile(model: dict[str, Any], ratings: list[dict[str, Any]] | None) -> dict[str, Any]:
    by_id: dict[str, dict[str, Any]] = {}
    for r in ratings or []:
        if r and r.get("competency_id"):
            by_id[r["competency_id"]] = r

    prior_mean = prior_mean_for_seniority(model["seniority"])
    competencies: list[dict[str, Any]] = []

    for c in model["competencies"]:
        r = by_id.get(c["id"], {})
        level = r.get("bars_level")
        level = float(level) if isinstance(level, (int, float)) else float("nan")
        confidence = _clamp(float(r.get("confidence") or 0), 0, 1)
        min_ep = (c.get("evidence_requirements") or {}).get("min_episodes", 1)
        episodes = _episodes(r.get("evidence"))

        import math
        insufficient = (not math.isfinite(level)) or r.get("state") == "unknown" or confidence < 0.35
        if not insufficient and episodes < min_ep:
            confidence = min(confidence, 0.55)
        state = "unknown" if insufficient else "measured"
        thin = state == "measured" and episodes < min_ep
        failure = _as_list(r.get("failure_triggered"))

        post: Posterior | None = None
        if state == "measured":
            eff_obs = confidence * _clamp(episodes, 1, 3)
            post = grm_posterior(level=int(_clamp(level, 1, 5)), weight=eff_obs, prior_mean=prior_mean)

        competencies.append({
            "id": c["id"], "name": c["name"], "cluster": c["cluster"], "weight": c["weight"],
            "state": state,
            "bars_level": int(_clamp(level, 1, 5)) if state == "measured" else None,
            "score": post.percentile if post else None,
            "theta": post.theta if post else None,
            "se": post.se if post else None,
            "ci90": list(post.ci90) if post else None,
            "ci_pct": _ci_pct(post) if post else None,
            "reliability": post.reliability if post else None,
            "confidence": confidence,
            "evidence": r.get("evidence", [])[:3] if isinstance(r.get("evidence"), list) else [],
            "followup": (r.get("followup") or _default_followup(c)) if state == "unknown" else None,
            "thin": thin,
            "risk": failure,
        })

    return _assemble(model, competencies, {"rater_count": 1})


def finalize_profile_ensemble(
    model: dict[str, Any],
    by_comp: dict[str, list[dict[str, Any]]],
    refutations: dict[str, Any] | None = None,
    meta: dict[str, Any] | None = None,
) -> dict[str, Any]:
    refutations = refutations or {}
    meta = meta or {}
    prior_mean = prior_mean_for_seniority(model["seniority"])
    n_raters = meta.get("rater_count", 1)
    import math

    competencies: list[dict[str, Any]] = []
    for c in model["competencies"]:
        raters = by_comp.get(c["id"], [])
        if not isinstance(raters, list):
            raters = []
        measured_raters = [
            r for r in raters
            if r and r.get("state") != "unknown"
            and isinstance(r.get("bars_level"), (int, float)) and math.isfinite(float(r["bars_level"]))
            and (float(r.get("confidence") or 0)) >= 0.35
        ]
        min_ep = (c.get("evidence_requirements") or {}).get("min_episodes", 1)
        refuted = bool(refutations.get(c["id"], {}).get("refuted"))

        if len(measured_raters) == 0 or len(measured_raters) < math.ceil(len(raters) / 2):
            fu = next((r["followup"] for r in raters if r and r.get("followup")), None) or _default_followup(c)
            competencies.append({
                "id": c["id"], "name": c["name"], "cluster": c["cluster"], "weight": c["weight"],
                "state": "unknown", "followup": fu, "rater_count": len(raters),
            })
            continue

        levels = [int(_clamp(float(r["bars_level"]), 1, 5)) for r in measured_raters]
        agreement = rater_agreement(levels)
        episodes_each = [_episodes(r.get("evidence")) for r in measured_raters]
        total_episodes = max(*episodes_each, 1) if episodes_each else 1
        thin = total_episodes < min_ep
        refute_factor = 0.4 if refuted else 1.0

        observations = []
        for i, r in enumerate(measured_raters):
            conf = _clamp(float(r.get("confidence") or 0), 0, 1)
            if thin:
                conf = min(conf, 0.55)
            observations.append({
                "level": int(_clamp(float(r["bars_level"]), 1, 5)),
                "weight": conf * _clamp(episodes_each[i], 1, 3) * refute_factor,
            })

        post = grm_posterior_multi(observations, prior_mean=prior_mean)
        assert post is not None
        mean_level = _js_round0(sum(levels) / len(levels))
        mean_conf = _clamp(sum(float(r.get("confidence") or 0) for r in measured_raters) / len(measured_raters), 0, 1) * refute_factor
        richest = sorted(measured_raters, key=lambda r: len(r.get("evidence") or []), reverse=True)[0]
        failures = [f for r in measured_raters for f in _as_list(r.get("failure_triggered"))]

        competencies.append({
            "id": c["id"], "name": c["name"], "cluster": c["cluster"], "weight": c["weight"],
            "state": "measured", "bars_level": mean_level,
            "score": post.percentile, "theta": post.theta, "se": post.se, "ci90": list(post.ci90),
            "ci_pct": _ci_pct(post), "reliability": post.reliability,
            "confidence": _js_round2(mean_conf),
            "rater_count": len(measured_raters),
            "rater_agreement": agreement,
            "refuted": refuted, "refute_reason": refutations.get(c["id"], {}).get("reason") if refuted else None,
            "evidence": (richest.get("evidence") or [])[:3] if isinstance(richest.get("evidence"), list) else [],
            "thin": thin, "risk": failures, "_levels": levels,
        })

    return _assemble(model, competencies, {**meta, "rater_count": n_raters})


def _assemble(model: dict[str, Any], competencies: list[dict[str, Any]], meta: dict[str, Any]) -> dict[str, Any]:
    measured = [c for c in competencies if c["state"] == "measured"]
    comp = composite_posterior([{"theta": c["theta"], "se": c["se"], "weight": c["weight"]} for c in measured])
    covered = sum(c["weight"] for c in measured)
    total_w = sum(c["weight"] for c in competencies)
    w_measured = covered or 0

    composite = comp.percentile if comp else None
    coverage = _js_round2(covered / total_w) if total_w > 0 else 0
    overall_confidence = _js_round2(sum(c["confidence"] * c["weight"] for c in measured) / w_measured) if w_measured > 0 else 0
    overall_reliability = _js_round2(sum((c.get("reliability") or 0) * c["weight"] for c in measured) / w_measured) if w_measured > 0 else 0

    dependability = None
    if meta.get("rater_count", 1) >= 2:
        with_levels = [c for c in measured if isinstance(c.get("_levels"), list) and len(c["_levels"]) >= 2]
        len_counts: dict[int, int] = {}
        for c in with_levels:
            n = len(c["_levels"])
            len_counts[n] = len_counts.get(n, 0) + 1
        if len_counts:
            target_len = sorted(len_counts.keys(), key=lambda k: len_counts[k], reverse=True)[0]
            matrix = [c["_levels"] for c in with_levels if len(c["_levels"]) == target_len]
            dependability = profile_dependability(matrix)
    for c in competencies:
        c.pop("_levels", None)

    critical_failures = [c["name"] for c in competencies if c.get("risk")]
    decision = decide_hire(
        theta=comp.theta if comp else None,
        se=comp.se if comp else None,
        coverage=coverage, critical_failures=critical_failures, seniority=model["seniority"],
    )
    if decision.band in ("Hire with Reservations", "Borderline"):
        decision.reservations = [
            c["name"] for c in competencies
            if c["state"] == "measured" and c["weight"] >= 0.08 and (c["score"] or 0) < 50
        ]

    return {
        "framework_version": model["framework_version"],
        "role_family": model["role_family"],
        "seniority": model["seniority"],
        "measurement_model": "grm-eap-v1",
        "rater_count": meta.get("rater_count", 1),
        "rater_dependability": dependability.dependability if dependability else None,
        "variance_components": (
            {"comp": dependability.v_comp, "rater": dependability.v_rater, "resid": dependability.v_resid}
            if dependability else None
        ),
        "competencies": competencies,
        "composite_score": composite,
        "composite_theta": comp.theta if comp else None,
        "composite_ci90": list(comp.ci90) if comp else None,
        "coverage": coverage,
        "overall_confidence": overall_confidence,
        "overall_reliability": overall_reliability,
        "decision": _decision_to_dict(decision),
        "unknown_competencies": [c["name"] for c in competencies if c["state"] == "unknown"],
        "followups": [{"competency": c["name"], "question": c["followup"]} for c in competencies if c.get("followup")],
    }


def _js_round0(n: float) -> int:
    import math
    return int(math.floor(n + 0.5))
